---
title: "lncRNAs_GitHub"
author: "Anja Hartewig"
date: "2023-05-04"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## setting up the environment
```{r}
library(rtracklayer)
library(DESeq2)
library(amap)
library(gplots)
library(RColorBrewer)
library(pheatmap)
library(ComplexHeatmap)
library(openxlsx)
library(dplyr)
library(circlize)
library(org.Hs.eg.db)
library(rstatix)
library(ggplot2)
library(gridExtra)
library(ggpubr) # for adding the stat_pvalue_manual

input_dir <- "C:/Users/gcanha/OneDrive - TUNI.fi/Documents/Secondary GBM/RNA-seq/"
output_dir <- "C:/Users/gcanha/OneDrive - TUNI.fi/Documents/Secondary GBM/microRNA Paper/"
miRNA_input <- "C:/Users/gcanha/OneDrive - TUNI.fi/Documents/Secondary GBM/microRNAs/"

writeLines(capture.output(sessionInfo()), paste0(output_dir, "lncRNAs_sessionInfo.txt"))
```

## extracting the lncRNAs from the DESeq2 data 
```{r}
RNA_results <- read.delim(paste0(input_dir, "astro_DESeq2_padj_included.CSV"),
                          header = TRUE, sep = ";")

## filter for padj < 0.05 and |LFC| > 1
RNA_sig_genes <- subset(RNA_results, RNA_results[, 7] < 0.05)
RNA_sig_genes <- subset(RNA_sig_genes, abs(RNA_sig_genes[, 3]) > 1)

RNA_counts <- read.delim(paste0(input_dir, "RNA_seq_gene_counts.txt"),
                         header = TRUE, sep = "\t")
```

## count the amount of lncRNAs based on the gencode annotation
Download from https://www.gencodegenes.org/human/release_30.html (Utilizing release 30), download of Long non-coding RNA gene annotation file
```{r}
# read in the gtf file and extract all gene names from the last field
all_lncRNAs <- rtracklayer::import("C:/Users/gcanha/OneDrive - TUNI.fi/Documents/Secondary GBM/gencode.v30.long_noncoding_RNAs.gtf")
all_lncRNAs_df <- as.data.frame(all_lncRNAs)

## write to table to export it to R version 3.6.
write.table(all_lncRNAs_df, file = paste0(input_dir, "GENCODE_v30_lncRNAs.txt"))

## better to put it into a dataframe to not have to convert it from a character to a vector
lncRNA_names <- as.data.frame(unique(all_lncRNAs_df[, c("gene_name", "gene_id")]))

## extracting all lncRNAs that are in the DESEq2 object
all_lncRNAs_DESeq2 <- RNA_results[(RNA_results[, 1] %in% lncRNA_names[, 1]), ]
all_lncRNAs_counts <- RNA_counts[(RNA_counts[, 1] %in% lncRNA_names[, 1]), ]

## export them to include in one of the other notebooks
write.table(all_lncRNAs_DESeq2, file = paste0(output_dir, "all_lncRNAs_DESeq2.txt"),
            sep = "\t", quote = FALSE)

## check how many of them in total are significant
sig_lncRNAs <- all_lncRNAs_DESeq2[which(all_lncRNAs_DESeq2[, 7] < 0.05), ]
sig_lncRNAs <- sig_lncRNAs[which(abs(sig_lncRNAs$log2FoldChange) > 1), ]

write.table(sig_lncRNAs, file = paste0(output_dir, "all_significant_lncRNAs_new.txt"),
            sep = "\t", row.names = FALSE, quote = FALSE)

# calculate the case-wise LFC for these lncRNAs
sig_lncRNAs_counts <- as.matrix(all_lncRNAs_counts[all_lncRNAs_counts[, 1] %in% sig_lncRNAs[, 1], 1:13])
rownames(sig_lncRNAs_counts) <- sig_lncRNAs_counts[, 1]
sig_lncRNAs_counts <- sig_lncRNAs_counts[, -(1)]

## write the normalized counts to a file
write.table(sig_lncRNAs_counts, file = paste0(output_dir, "norm_counts_all_significant_lncRNAs.txt"),
            sep = "\t", row.names = TRUE, quote = FALSE)

## heatmap was produced in "RNA_seq_analysis_GitHub_updated.Rmd"
```

## looking at the correlation between the predicted targets of the lncRNAs 
The targets were downloaded 05.05.23 from http://ncpath.pianlab.cn/#/Home (NcPath)
Additional targets were downloaded from LncTarD (https://lnctard.bio-database.com/) on 25.05.23
```{r}
# reading in all the targets
path_NcPath <- "C:/Users/gcanha/OneDrive - TUNI.fi/Documents/Secondary GBM/lncRNAs/NcPath_targets/"
path_LncTarD <- "C:/Users/gcanha/OneDrive - TUNI.fi/Documents/Secondary GBM/lncRNAs/LncTarD_targets/"

filenames_NcPath <- list.files(path_NcPath,
                               pattern = "*.xlsx", full.names = TRUE)

filenames_LncTarD <- list.files(path_LncTarD,
                                pattern = "*.txt", full.names = TRUE)

all_lncRNA_targets_NcPath <- lapply(filenames_NcPath, read.xlsx)
all_lncRNA_targets_LncTarD <- lapply(filenames_LncTarD, read.delim, sep = ",")

# obtaining the sample names from the filenames
lncRNA_names <- sub("^.+/", "", filenames_NcPath)
lncRNA_names <- substr(lncRNA_names, 1, nchar(lncRNA_names) - 5)
names(all_lncRNA_targets_NcPath) <- lncRNA_names
# Note: filenames for LncTarD files are different from NcPath
lncRNA_names <- sub("^.+/", "", filenames_LncTarD)
lncRNA_names <- substr(lncRNA_names, 1, nchar(lncRNA_names) - 4)
names(all_lncRNA_targets_LncTarD) <- lncRNA_names

# extract unique targets
# make a table with lncRNA + target to calculate correlations
# the name appearing more than once in the list means it only has one prediction
all_lncRNA_targets_NcPath_filtered <- list()
for (i in 1:length(all_lncRNA_targets_NcPath)){
  all_lncRNA_targets_NcPath_filtered[[i]] <- all_lncRNA_targets_NcPath[[i]] %>% distinct(TargetGene, .keep_all = TRUE)
}

# repeat for LncTarD targets
all_lncRNA_targets_LncTarD_filtered <- list()
for (i in 1:length(all_lncRNA_targets_LncTarD)){
  all_lncRNA_targets_LncTarD_filtered[[i]] <- all_lncRNA_targets_LncTarD[[i]] %>% distinct(Target, .keep_all = TRUE)
}

## extracting the TargetGene and LncRNA columns from each dataframe
LncRNA_target_df <- as.data.frame(matrix(ncol = 2, nrow = 1))
colnames(LncRNA_target_df) <- c("LncRNA", "TargetGene")
for (i in 1:length(all_lncRNA_targets_NcPath_filtered)) {
  temp_df <- c()
  temp_LncRNA <- all_lncRNA_targets_NcPath_filtered[[i]]$LncRNA
  temp_TargetGene <- all_lncRNA_targets_NcPath_filtered[[i]]$TargetGene
  temp_df <- cbind(temp_LncRNA, temp_TargetGene)
  colnames(temp_df) <- c("LncRNA", "TargetGene")
  LncRNA_target_df <- rbind(LncRNA_target_df, temp_df)
}
# remove the first row with NA in them
LncRNA_target_df <- LncRNA_target_df[-(1), ]

## repeat for LncTarD targets
LncRNA_target_df_2 <- as.data.frame(matrix(ncol = 2, nrow = 1))
colnames(LncRNA_target_df_2) <- c("Regulator", "Target")
for (i in 1:length(all_lncRNA_targets_LncTarD_filtered)) {
  temp_df <- c()
  temp_Regulator <- all_lncRNA_targets_LncTarD_filtered[[i]]$Regulator
  temp_Target <- all_lncRNA_targets_LncTarD_filtered[[i]]$Target
  temp_df <- cbind(temp_Regulator, temp_Target)
  colnames(temp_df) <- c("Regulator", "Target")
  LncRNA_target_df_2 <- rbind(LncRNA_target_df_2, temp_df)
}
# remove the first rows with NA in them
LncRNA_target_df_2 <- LncRNA_target_df_2[-(1), ]

## combine them into one dataframe and then remove any duplicates
## adjust the column names for combining
colnames(LncRNA_target_df_2) <- c("LncRNA", "TargetGene")
LncRNA_target_df_combined <- rbind(LncRNA_target_df, LncRNA_target_df_2)

## look if there are any duplicates
LncRNA_target_df_combined <- unique(LncRNA_target_df_combined)
```

## calculating the correlation
This code is implemented similarly to the miRNA target correlation analysis (target_prediction_miRNA_mRNA.Rmd)
```{r}
## get the case-wise normalized RNA counts
RNA_counts <- read.delim(paste0(input_dir, "norm_counts_RNA-seq.csv"),
                                    header = TRUE, sep = ";")

RNA_results <- read.delim(paste0(input_dir, "astro_DESeq2_padj_included.CSV"),
                           header = TRUE, sep = ";")

## make a filtered object for combining with the miRNAs
RNA_counts_filtered <- RNA_counts[RNA_counts[, 1] %in% RNA_results[, 1], ]

## read in the data for all miRNAs with adj. p-val
miRNA_results <- read.delim(paste0(miRNA_input, "miRNAs_DESeq2_only_R.csv"),
                           header = TRUE, sep = ",")
## remove all those that have #N/A in their adjusted p-val
miRNA_results_filtered <- miRNA_results[-which(miRNA_results[, 7] == "#N/A"), ]

norm_miRNA_counts <- read.delim(paste0(miRNA_input, "miRNA_normalized_R.csv"),
                                      header = TRUE, sep = ";")
norm_miRNA_counts_filtered <- norm_miRNA_counts[norm_miRNA_counts[, 1] %in% miRNA_results_filtered[, 1], ]

# the format of target miRNA is e.g. "miR-124", change the counts accordingly
norm_miRNA_counts_filtered[, 1] <- substr(norm_miRNA_counts_filtered[, 1], 5, nchar(norm_miRNA_counts_filtered[, 1]))
miRNA_results_filtered[, 1] <- substr(miRNA_results_filtered[, 1], 5, nchar(miRNA_results_filtered[, 1]))

## make a combined object to run only 1 correlation analysis
combined_filtered_counts <- rbind(RNA_counts_filtered, norm_miRNA_counts_filtered)

## check if all the targets have an adjusted p-value
no_expression_data <- setdiff(LncRNA_target_df_combined$TargetGene, combined_filtered_counts[, 1])
RNA_lncRNA_corr <- LncRNA_target_df_combined[LncRNA_target_df_combined$TargetGene %in% combined_filtered_counts[, 1], ]

## there are genes such as "circPVT1" LncRNA since they are regulating a ncRNA
## filter them out
RNA_lncRNA_corr <- RNA_lncRNA_corr[RNA_lncRNA_corr$LncRNA %in% combined_filtered_counts[, 1], ]

## format the counts to have the gene names as rownames
rownames(combined_filtered_counts) <- combined_filtered_counts[, 1]
combined_filtered_counts <- combined_filtered_counts[, (-1)]

## calculate the correlation
correlations <- RNA_lncRNA_corr
correlations_norm <- matrix(ncol = 3, nrow = dim(RNA_lncRNA_corr)[1])
for (i in 1:dim(correlations)[1]){
  correlations_norm[i, 1] <- cor(as.numeric(combined_filtered_counts[correlations[i, 1], ]),
                                 as.numeric(combined_filtered_counts[correlations[i, 2], ]))
  correlations_norm[i, 2] <- cor.test(as.numeric(combined_filtered_counts[correlations[i, 1], ]),
                                      as.numeric(combined_filtered_counts[correlations[i, 2], ]))$p.value
  correlations_norm[i, 3] <- cor(as.numeric(combined_filtered_counts[correlations[i, 1], ]),
                                 as.numeric(combined_filtered_counts[correlations[i, 2], ]), method = "spearman")
}

correlations <- as.data.frame(cbind(correlations, correlations_norm))
colnames(correlations) <- c("lncRNA", "target", "pearson correlation",
                            "pearson correlation p-val",
                            "spearman correlation")

## extract the spearman correlation for the score in the GRNs
write.table(correlations, file = paste0(output_dir, "all_correlations_unfiltered.txt"),
            quote = FALSE, row.names = FALSE)

## read the correlations back in
correlations <- read.table(paste0(output_dir, "all_correlations_unfiltered.txt"),
                           sep = " ", header = TRUE)

spearman_correlation_scores <- correlations[, c(1:2, 5)]
write.table(spearman_correlation_scores, file = paste0(output_dir, "spearman_correlations_unfiltered.txt"),
            quote = FALSE, row.names = FALSE)

## filter for correlating targets
correlations_filtered <- correlations[which(as.numeric(correlations[, 4]) < 0.05), ]
## for now include both positive and negative correlating genes
positive_correlation <- c()
negative_correlation <- c()
for (i in 1:dim(correlations_filtered)[1]){
  positive_correlation[i] <- as.numeric(correlations_filtered[i, 3]) > 0.5 && as.numeric(correlations_filtered[i, 5]) > 0.5
  negative_correlation[i] <- as.numeric(correlations_filtered[i, 3]) < -0.5 && as.numeric(correlations_filtered[i, 5]) < -0.5
}

combined_correlation <- rbind(correlations_filtered[positive_correlation, ],
                              correlations_filtered[negative_correlation, ])

## extract some statistics for the correlating targets
## HAGLROS is filtered out later on (not relevant)
RNA_results_corr_targets <- RNA_results[RNA_results[, 1] %in% combined_correlation[, 2], ]
miRNA_results_corr_targets <- miRNA_results_filtered[miRNA_results_filtered[, 1] %in% combined_correlation[, 2], ]
colnames(miRNA_results_corr_targets)[1] <- "genes"
statistics_corr_targets <- rbind(RNA_results_corr_targets, miRNA_results_corr_targets)

## combine with the correlation
RNA_stat_lncRNA_targets <- as.data.frame(matrix(ncol = 7, nrow = dim(combined_correlation[1])))
for (i in 1:dim(combined_correlation)[1]) {
  index <- grep(combined_correlation[i, 2], statistics_corr_targets[, 1])
  RNA_stat_lncRNA_targets[i, ] <- statistics_corr_targets[index, ]
}
colnames(RNA_stat_lncRNA_targets) <- colnames(RNA_results)

combined_correlation <- cbind(combined_correlation, RNA_stat_lncRNA_targets)

## filter for those for which no statistic is available
combined_correlation <- combined_correlation[which(combined_correlation$target == combined_correlation$genes), ]

## make sure the columns are numeric before filtering
combined_correlation[, 12] <- as.numeric(combined_correlation[, 12])

# filter for DE genes
DE_target_genes <- combined_correlation[combined_correlation[, 12] < 0.05, ]

## remove the WDFY3-AS2 due to the cut-off
DE_target_genes <- DE_target_genes[-which(DE_target_genes$lncRNA == "WDFY3-AS2"), ]

write.table(unique(DE_target_genes),
            file = paste0(output_dir, "all_correlating_DE_target_genes_lncRNAs_stats.txt"),
            quote = FALSE, row.names = FALSE, col.names = TRUE)


## check how many of those have |LFC| > 1
DE_target_genes_FC_1 <- DE_target_genes[abs(DE_target_genes[, 8]) > 1, ]

## write the gene names for both of them to a file for enrichr
write.table(unique(DE_target_genes_FC_1[, 2]),
            file = paste0(output_dir, "all_correlating_DE_target_genes_lncRNAs_FC_1.txt"),
            quote = FALSE, row.names = FALSE, col.names = FALSE)

table(DE_target_genes_FC_1[, 2])[table(DE_target_genes_FC_1[, 2]) > 1]

## check again for how many lncRNAs we have correlating targets
write.table(table(DE_target_genes[, 1]), file = paste0(output_dir, "Nr._correlating_DE_lncRNAs_cohorts.tsv"),
            quote = FALSE, row.names = FALSE, sep = "\t")

## read in the miRNA targets for intersection
miRNA_targets <- read.delim(paste0(miRNA_input, "all_correlating_DE_target_genes_miRNAs_FC_1.txt"),
                            sep = "\t", header = FALSE)

length(intersect(miRNA_targets[, 1], DE_target_genes_FC_1[, 2]))
```

## make a heatmap for the 62 DE lncRNA targets (SFig 2a)
```{r}
## DE_target_genes, get the counts for them
DE_target_gene_counts <- RNA_counts[(RNA_counts[, 1] %in% DE_target_genes[, 2]), ]

## obtain the LFCs as well
DE_target_LFC <- RNA_results[(RNA_results[, 1] %in% DE_target_genes[, 2]), ]
## make sure the order is the same
DE_target_gene_counts_sorted <- DE_target_gene_counts[order(DE_target_gene_counts[, 1]), ]
DE_target_LFC_sorted <- DE_target_LFC[order(DE_target_LFC[, 1]), ]

sum(DE_target_gene_counts_sorted[, 1] == DE_target_LFC_sorted[, 1])

rownames(DE_target_gene_counts_sorted) <-DE_target_gene_counts_sorted[, 1]
DE_target_gene_counts_sorted <- DE_target_gene_counts_sorted[, -(1)]

colnames(DE_target_gene_counts_sorted) <- c("TG02a", "TG02b", "TG03a", "TG03b",
                                     "TG04a", "TG04b", "TG05a", "TG05b",
                                     "TG01a", "TG01b", "TG06a", "TG06b")

DE_target_gene_counts_sorted_log <- log2((DE_target_gene_counts_sorted + 1))

grade <- as.factor(as.numeric(c(3, 4, 3, 4, 3, 4, 2, 4, 3, 4, 2, 4)))

ha <- HeatmapAnnotation(grade = grade,
                        col = list(grade = c("2" = "#ffae00b2",
                                             "3" = "#775c40b2",
                                             "4" = "#89138dcb")))

heatmap_pal <- colorRamp2(c(-4, 0, 4), c("#3C5488B2", "white", "#E64B35B2"))

## make the rowannotation
## make a separate data.frame containing the information
lncRNAs_anno <- unique(DE_target_genes[, 1])

## loop through the lncRNAs
lncRNA_anno <- as.data.frame(matrix(data = "none", ncol = 11,
                                    nrow = nrow(DE_target_gene_counts_sorted_log)))
colnames(lncRNA_anno) <- lncRNAs_anno
rownames(lncRNA_anno) <- rownames(DE_target_gene_counts_sorted_log)
for (i in seq_along(lncRNAs_anno)) {
  lncRNA <- lncRNAs_anno[i]
  corr_targets <- DE_target_genes[DE_target_genes[, 1] == lncRNA, ]
  pos_corr_targets <- corr_targets[corr_targets[, 3] > 0, ]
  neg_corr_targets <- corr_targets[corr_targets[, 3] < 0, ]
  if (length(pos_corr_targets) > 0) {
    for (j in seq_along(pos_corr_targets[, 1])) {
    indices <- which(rownames(lncRNA_anno) == pos_corr_targets[j, 2])
    lncRNA_anno[indices, i] <- "positive"
    }
  }
  if (length(neg_corr_targets) > 0) {
    for (j in seq_along(neg_corr_targets[, 1])) {
    indices <- which(rownames(lncRNA_anno) == neg_corr_targets[j, 2])
    lncRNA_anno[indices, i] <- "negative"
    }
  }
}

# Define the colors for the annotations as a named vector
annotation_colors <- c("none" = "white", "positive" = "#dd7b7b", "negative" = "#177a176e")
LFC_col <- colorRamp2(c(-4, 0, 4), c("#00b7d7", "#d5d3d3", "#f83479"))

# Create the row annotation
row_annotation <- rowAnnotation(
  target_of_COLCA1 = anno_simple(lncRNA_anno[, 1], col = annotation_colors),
  target_of_CYTOR = anno_simple(lncRNA_anno[, 2], col = annotation_colors),
  target_of_FBXL19_AS1 = anno_simple(lncRNA_anno[, 3], col = annotation_colors),
  target_of_HAGLR = anno_simple(lncRNA_anno[, 4], col = annotation_colors),
  target_of_LINC01088 = anno_simple(lncRNA_anno[, 5], col = annotation_colors),
  target_of_MIR4435_2HG = anno_simple(lncRNA_anno[, 6], col = annotation_colors),
  target_of_NEAT1 = anno_simple(lncRNA_anno[, 7], col = annotation_colors),
  target_of_NRG3_AS1 = anno_simple(lncRNA_anno[, 8], col = annotation_colors),
  target_of_PVT1 = anno_simple(lncRNA_anno[, 9], col = annotation_colors),
  target_of_RMST = anno_simple(lncRNA_anno[, 10], col = annotation_colors),
  target_of_CRNDE = anno_simple(lncRNA_anno[, 11], col = annotation_colors),
  LFC = anno_simple(DE_target_LFC_sorted[, 3], col = LFC_col)
)

annotation_legend_1 <- Legend(
  labels = names(annotation_colors),
  legend_gp = gpar(fill = annotation_colors),
  title = "Correlation"
)

annotation_legend_2 <- Legend(col_fun = LFC_col,
                              title = "LFC",
                              at = c(-4, 0, 4),
                              labels = c("-4", "0", "4"))

DE_target_gene_counts_sorted_log_scaled <- t(scale(t(DE_target_gene_counts_sorted_log)))
## order them alphabetically
ordered_indices <- c(9, 1, 3, 5, 7, 11, 10, 2, 4, 6, 8, 12)
DE_target_gene_counts_sorted_log_scaled <- DE_target_gene_counts_sorted_log_scaled[, ordered_indices]


pdf(file = paste0(output_dir, "Heatmap_DE_targets_log2_scaled_pearson_new.pdf"), width = 12, height = 10)
ht <- Heatmap(DE_target_gene_counts_sorted_log_scaled, top_annotation = ha,
        left_annotation = row_annotation,
        name = "z-score", clustering_distance_rows = "pearson",
        column_title = "log2 counts of all DE lncRNA targets, scaled",
        col = heatmap_pal, cluster_columns = FALSE,
        row_names_gp = gpar(fontsize = 9), cluster_rows = FALSE)
draw(ht, annotation_legend_side = "right", annotation_legend_list = list(annotation_legend_1, annotation_legend_2))
dev.off()
```

## calculcate the correlation between let7b and NEAT1
```{r}
NEAT1_counts <- RNA_counts[RNA_counts[, 1] == "NEAT1", ]
rownames(NEAT1_counts) <- NEAT1_counts[, 1]
NEAT1_counts <- NEAT1_counts[, -(1)]
let7b_counts <- norm_miRNA_counts[norm_miRNA_counts[, 1] == "hsa-let-7b-3p", ]
rownames(let7b_counts) <- let7b_counts[, 1]
let7b_counts <- let7b_counts[, -(1)]

cor(as.numeric(NEAT1_counts[1, ]), as.numeric(let7b_counts[1, ]))
# cor.test(as.numeric(NEAT1_counts[1, ]), as.numeric(let7b_counts[1, ]))$p.value
cor(as.numeric(NEAT1_counts[1, ]), as.numeric(let7b_counts[1, ]), method = "spearman")
## pearson: -0.47 spearman: -0.65 (p-val not significant)

## check for some of the other let7 members
let7a_counts <- norm_miRNA_counts[norm_miRNA_counts[, 1] == "hsa-let-7a-3p", ]
rownames(let7a_counts) <- let7a_counts[, 1]
let7a_counts <- let7a_counts[, -(1)]

cor(as.numeric(NEAT1_counts[1, ]), as.numeric(let7a_counts[1, ]))
cor(as.numeric(NEAT1_counts[1, ]), as.numeric(let7a_counts[1, ]), method = "spearman")
## spearman: -0.82, -0.4 for pearson

## 7b 5p -0.44 pearson, -0.47 spearman

## correlation with e.g. let-7c-5p

let7a_counts <- norm_miRNA_counts[norm_miRNA_counts[, 1] == "hsa-let-7d-5p", ]
rownames(let7a_counts) <- let7a_counts[, 1]
let7a_counts <- let7a_counts[, -(1)]

cor(as.numeric(NEAT1_counts[1, ]), as.numeric(let7a_counts[1, ]))
cor(as.numeric(NEAT1_counts[1, ]), as.numeric(let7a_counts[1, ]), method = "spearman")
## spearman: -0.82, -0.4 for pearson
```

## look at how many of our DE lncRNAs can be found in TCGA as well
```{r}
## read in the TCGA data
TCGA_DE_mRNAs <- read.delim("C:/Users/gcanha/OneDrive - TUNI.fi/Documents/Secondary GBM/microRNA Paper/TCGA comparison/DE_mRNAs_TCGA.txt")
DE_lncRNAs_TG <- read.delim(file = paste0(output_dir, "all_significant_lncRNAs_q_01.txt"),
                            sep = "\t")
## note: this is only the 0.05 cutoff
test <- TCGA_DE_mRNAs[which(TCGA_DE_mRNAs[, 2] %in% DE_lncRNAs_TG[, 1]), ]

## read in the Gencode to check how many DElncRNAs are there in total
all_lncRNAs <- rtracklayer::import("C:/Users/gcanha/OneDrive - TUNI.fi/Documents/Secondary GBM/gencode.v30.long_noncoding_RNAs.gtf")
all_lncRNAs_df <- as.data.frame(all_lncRNAs)

DE_lncRNAs_TCGA <- TCGA_DE_mRNAs[which(TCGA_DE_mRNAs[, 2] %in% all_lncRNAs_df[, 12]), ]
```

## read in the negatively correlating miRNA targets and look for overlaps
```{r}
## read in the lncRNA targets
lncRNA_corr_targets <- read.table(file = paste0(output_dir, "lncRNA_corr_targets.txt"),
                                  sep = "\t", header = TRUE)

## read in miRNA targets
miRNA_targets <- read.xlsx("C:/Users/gcanha/OneDrive - TUNI.fi/Documents/Secondary GBM/microRNAs/target_genes_FC_0.3_no_30a.xlsx")
## the first two columns are the most important ones

## try to find overlaps
co_regulated_targets <- intersect(lncRNA_corr_targets[, 2], miRNA_targets[, 2])
```

## filtering the annotations obtained from NcPath
Starting from the unfiltered all_lncRNA_targets_NcPath object read in above
```{r}
path_NcPath <- "C:/Users/gcanha/OneDrive - TUNI.fi/Documents/Secondary GBM/lncRNAs/NcPath_targets/"

filenames_NcPath <- list.files(path_NcPath,
                               pattern = "*.xlsx", full.names = TRUE)

all_lncRNA_targets_NcPath <- lapply(filenames_NcPath, read.xlsx)

# obtaining the sample names from the filenames
lncRNA_names <- sub("^.+/", "", filenames_NcPath)
lncRNA_names <- substr(lncRNA_names, 1, nchar(lncRNA_names) - 5)
names(all_lncRNA_targets_NcPath) <- lncRNA_names

# not filtering for unique targets (to see the different KEGG pathways)
## extracting the TargetGene and LncRNA and hsaid columns from each dataframe
LncRNA_pathway_df <- as.data.frame(matrix(ncol = 3, nrow = 1))
colnames(LncRNA_pathway_df) <- c("LncRNA", "TargetGene", "PathwayID")
for (i in 1:length(all_lncRNA_targets_NcPath)) {
  temp_df <- c()
  temp_LncRNA <- all_lncRNA_targets_NcPath[[i]]$LncRNA
  temp_TargetGene <- all_lncRNA_targets_NcPath[[i]]$TargetGene
  temp_pathway_id <- all_lncRNA_targets_NcPath[[i]]$hsaid
  temp_df <- cbind(temp_LncRNA, temp_TargetGene, temp_pathway_id)
  colnames(temp_df) <- c("LncRNA", "TargetGene", "PathwayID")
  LncRNA_pathway_df <- rbind(LncRNA_pathway_df, temp_df)
}
# remove the first row with NA in them
LncRNA_pathway_df <- LncRNA_pathway_df[-(1), ]

## filter for the targets we know are correlating
# the %in% doesn't work since targets that are correlating in at least 1 lncRNA will still remain
# make a for-loop instead
correlating_targets_pathways <- c()
# loop through all the lncRNAs
for (i in 1:length(unique(LncRNA_pathway_df$LncRNA))){
  LncRNA <- unique(LncRNA_pathway_df$LncRNA)[i]
  predicted_targets <- LncRNA_pathway_df[which(LncRNA_pathway_df[, 1] == LncRNA), ]
  corr_targets <- lncRNA_corr_targets[which(lncRNA_corr_targets[, 1] == LncRNA), ]
  temp <- predicted_targets[predicted_targets[, 2] %in% corr_targets[, 2], ]
  correlating_targets_pathways <- rbind(correlating_targets_pathways, temp)
}


# remove the duplicates
correlating_targets_pathways <- correlating_targets_pathways[!duplicated(correlating_targets_pathways), ]

## look at what are the most frequent pathways
sort(table(correlating_targets_pathways$PathwayID))
## most targets are related to hsa05200 pathways in cancer
## a total of 211 pathways were detected
## filter out all pathways with less than 15 targets
Pathways_filtered <- as.data.frame(table(correlating_targets_pathways$PathwayID))
Pathways_filtered <- Pathways_filtered[which(Pathways_filtered$Freq > 15), ]
# 110 pathways are left with more than 10 targets, 64 with more than 15 targets
## sort for the most common ones
Pathways_filtered_sorted <- Pathways_filtered[order(Pathways_filtered$Freq, decreasing = TRUE), ]
```
